### PR TITLE
Rework websocket server

### PR DESCRIPTION
**Pull request by coolo**
The main goal is to detect dead workers purely on their websocket activity,
but instead of writing everything into the database all the time, just
keep the last_seen timestamp in a local hash

This is once again deep in not covered code, but I tested this heavily for
performance on an unnamed instance with > 100 workers

### CODE REVIEW

**Comment by coolo** [comment]
looks like I broke live log

File: lib/OpenQA/WebSockets/Server.pm [hunk 1]
```diff
@@ -287,7 +278,7 @@ sub setup {
     app->secrets(['nosecretshere']);
 
     # start worker checker - check workers each 2 minutes
-    Mojo::IOLoop->recurring(120 => \&_workers_checker);
+    Mojo::IOLoop->recurring(5 => \&_workers_checker);
```
Comment (coolo) [inline lib/OpenQA/WebSockets/Server.pm]: this is only for testing. no need to stress the database so much

**Comment by coolo** [comment]
but we can run now pretty reliable 112 jobs.But 230 seconds without websocket activity seems pretty normal. Not only at the start of the job, but also during heavy upload activity.

File: lib/OpenQA/WebSockets/Server.pm [hunk 2]
```diff
@@ -210,14 +192,24 @@ sub _get_stale_worker_jobs {
...
     my $schema = OpenQA::Schema::connect_db;
 
+    # grab the workers we've seen lately
+    my @ok_workers;
+    for my $worker (values %$workers) {
+        if (time - $worker->{last_seen} <= $threshold) {
+            push(@ok_workers, $worker->{id});
+        }
+        else {
+            log_debug(sprintf("Worker %s not seen since %d seconds", $worker->{db}->name, time - $worker->{last_seen}));
```
Comment (okurz) [inline lib/OpenQA/WebSockets/Server.pm]: sprintf shouldn't be needed

File: lib/OpenQA/WebSockets/Server.pm [hunk 3]
```diff
@@ -197,11 +179,11 @@ sub _message {
             $worker->set_property('STOP_WAITFORNEEDLE_REQUESTED', $prop->{waitforneedle} ? 1 : 0);
         }
         else {
-            $ws->app->log->error("Unknown property received from worker $workerid");
+            log_error("Unknown property received from worker $worker->{id}");
         }
     }
     else {
-        $ws->app->log->error(sprintf('Received unknown message type "%s" from worker %u', $json->{type}, $workerid));
+        log_error(sprintf('Received unknown message type "%s" from worker %u', $json->{type}, $worker->{id}));
```
Comment (okurz) [inline lib/OpenQA/WebSockets/Server.pm]: What's better about `log_error` compared to `$ws->app->log->error`?

Comment (coolo) [inline lib/OpenQA/WebSockets/Server.pm]: more consistent with log_debug and log_warning

**Comment by coolo** [comment]
issue with live log is sorted

File: lib/OpenQA/WebSockets/Server.pm [hunk 4]
```diff
@@ -112,70 +100,64 @@ sub check_authorized {
...
-    my ($self)   = @_;
+    my ($self) = @_;
     my $workerid = $self->param('workerid');
-    my $worker   = _validate_workerid($workerid);
-    unless ($worker) {
-        return $self->render(text => 'Unauthorized', status =>);
+    unless ($workers->{$workerid}) {
+        my $db = app->schema->resultset("Workers")->find($workerid);
+        unless ($db) {
+            return $self->render(text => 'Unauthorized', status =>);
```
Comment (aaannz) [inline lib/OpenQA/WebSockets/Server.pm]: `status =>` is intentional? I guess you're missing 401 or 403.

Comment (coolo) [inline lib/OpenQA/WebSockets/Server.pm]: I only indented this line. But I wondered about the same :)

Comment (aaannz) [inline lib/OpenQA/WebSockets/Server.pm]: Touché.
There is this thing with replying some weird data on websockets we once discussed on IRC though. Can't remember and my server is still down, so can't access IRC logs, but @kraih had some comments too.
If it is possible, I'm for 401.

Comment (coolo) [inline lib/OpenQA/WebSockets/Server.pm]: and your memory is about #903 - which is about success and not error

**Pull request merged**